### PR TITLE
Refactored BatchedTensorImpl to have single bdim and level

### LINKED ISSUE #86: Refactor BatchedTensorImpl
[BatchedTensorImpl](https://github.com/facebookresearch/functorch/blob/42a8f6608147fc41debdb3bc9d632dcba1be1a99/functorch/csrc/BatchedTensorImpl.h#L67) got copy-pasted from our previous vmap prototype. It handles multiple batch dimensions at the [same time](https://github.com/facebookresearch/functorch/blob/42a8f6608147fc41debdb3bc9d632dcba1be1a99/functorch/csrc/BatchedTensorImpl.h#L108).

functorch assumes that BatchedTensorImpl only has a single batch dimension. We should change the representation of BatchedTensorImpl to holding a Tensor value, an int64_t bdim, and an int64_t level.

This will:
- improve our codebase
- Make things faster because there is one less layer of indirection

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Refactored BatchedTensorImpl to have single bdim and level
- Made uniform signature to receive (level, bdim) as args. Method makeBatched is impacted in many files.
  - I was based on `addBatchDim` signature and maybe I could adopt (bdim, level) but both were defined.

Fixes #86

### CODE REVIEW

**Comment by vfdev-5** [comment]
@zou3519 any comments to this PR, please ?

File: functorch/csrc/BatchedTensorImpl.cpp [hunk 1]
```diff
@@ -175,9 +170,12 @@ bool inplaceIsVmapCompatible(const Tensor& self, const Tensor& other) {
...
   }
-  auto self_levels = createVmapLevelsBitset(self_batched->bdims());
-  auto other_levels = createVmapLevelsBitset(other_batched->bdims());
-  return self_levels == (self_levels | other_levels);
+  // TODO(vfdev): As BatchedTensorImpl is refactored and has only one dim and one level.
+  // Below code may be simplified.
+  // auto self_levels = createVmapLevelsBitset(self_batched->level());
+  // auto other_levels = createVmapLevelsBitset(other_batched->level());
+  // return self_levels == (self_levels | other_levels);
+  return self_batched->level() == other_batched->level();
```
Comment (zou3519) [inline functorch/csrc/BatchedTensorImpl.cpp:178]: I don't think this function is used anywhere actually. Could we delete it? (as well as the declaration in the header)

**Review by zou3519 — COMMENTED** [review]
Looks pretty good. I kind of prefer having (bdim, level) instead of (level, bdim). Let me know your thoughts

File: functorch/csrc/BatchRulesModules.cpp [hunk 2]
```diff
@@ -191,8 +191,8 @@ std::tuple<Tensor,Tensor> cudnn_convolution_backward_plumbing(const Tensor & sel
         padding, stride, dilation, groups,
         benchmark, deterministic, allow_tf32, output_mask);
     return std::make_tuple(
-        makeBatched(std::get<0>(result), std::get<1>(result), cur_level),
-        makeBatched(std::get<2>(result), std::get<3>(result), cur_level));
+        makeBatched(std::get<0>(result), cur_level, std::get<1>(result)),
+        makeBatched(std::get<2>(result), cur_level, std::get<3>(result)));
```
[the remark this answers is not in the document]
Comment (zou3519) [reply functorch/csrc/BatchRulesModules.cpp:195]: Yes, I think everywhere is fine

**Comment by vfdev-5** [comment]
@zou3519 I updated the code as you asked, please review

File: functorch/csrc/BatchedTensorImpl.h [hunk 3]
```diff
@@ -31,57 +31,38 @@ constexpr int64_t kVmapNumLevels = 64;
...
+//    bt = BatchedTensorImpl(ones(2, 3, 5, 7), lvl=1, dim=0)
+// dimension 0 is batch dimension.
 //
 // bt.sizes() returns (5, 7); bt.sum(0) performs a reduction over the (public)
 // dim 0, which is equivalent to dim 3 in the underlying ones(2, 3, 5, 7) tensor.
 struct BatchedTensorImpl : public c10::TensorImpl {
-  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
-  explicit BatchedTensorImpl(at::DispatchKeySet key_set, Tensor value, BatchDims bdims);
+  explicit BatchedTensorImpl(Tensor value, int64_t dim, int64_t level);
+  explicit BatchedTensorImpl(at::DispatchKeySet key_set, Tensor value, int64_t dim, int64_t level);
```
Comment (zou3519) [inline functorch/csrc/BatchedTensorImpl.h:47]: nit: these should be `bdim`, not `dim`, for clarity

**Review by zou3519 — APPROVED** [review]
Thanks! everything is so much nicer now

**Pull request merged**